ibis/common/grounds.py

```diff
@@ -131,24 +133,39 @@ def __repr__(self) -> str:
         argstring = ", ".join(args)
         return f"{self.__class__.__name__}({argstring})"
 
-    def __eq__(self, other):
+    def __eq__(self, other) -> bool:
         if type(self) is not type(other):
             return NotImplemented
 
         return all(
-            getattr(self, n) == getattr(other, n) for n in self.__argnames__
+            getattr(self, n, None) == getattr(other, n, None)
+            for n in self.__attributes__.keys()
         )
 
-    def __getstate__(self):
-        return {name: getattr(self, name) for name in self.__argnames__}
+    def __getstate__(self) -> dict[str, Any]:
+        return {n: getattr(self, n, None) for n in self.__attributes__.keys()}
 
-    def __setstate__(self, state):
+    def __setstate__(self, state) -> None:
         self.__init__(**state)
 
-    def copy(self, **overrides):
-        kwargs = self.__getstate__()
-        kwargs.update(overrides)
-        return self.__class__(**kwargs)
+    def copy(self, **overrides) -> Annotable:
+        """Return a copy of this object with the given overrides.
+
+        Parameters
+        ----------
+        overrides : dict
+
+        Returns
+        -------
+        copy : Annotable
+        """
+        this = copy(self)
+        for name, value in overrides.items():
+            if field := self.__attributes__.get(name):
```

Comment:
Should this check for `not None` instead of falsey?

```suggestion
            if (field := self.__attributes__.get(name)) is not None:
```